README.md
# [Bitcointip](http://www.reddit.com/r/bitcointip) user script

This is a Greasemonkey user script that adds a "tip bitcoins" button
next to reddit's "give gold". It also displays
[bitcointip's](http://www.reddit.com/u/bitcointip/) verifications more
concisely. Here's what it looks like.

![](http://i.imgur.com/BHd96.png)

![](http://i.imgur.com/HseZA.png)

It's verified to work in both Firefox with Greasemonkey and Chrome
with Tampermonkey. It should work in other browsers that support user
scripts, as long as they support `unsafeWindow`.

## Origin

This script was born
[here](http://www.reddit.com/r/Bitcoin/comments/14ns5n/why_not_a_tip_bitcoin_link_on_reddit_i_was/c7f04ay). Send
complaints, requests, bug reports, and bitcoin tips to
[skeeto](http://www.reddit.com/u/skeeto) (me) on reddit (or make an
issue here).


changelog.md
 * 2.1.0
   * Merge in changes/fixes from RES
   * Subreddit-enabled icon off by default
 * 2.0.2
   * Support for new +/u/bitcointip keyword
   * Small fixes
 * 2.0.0
   * Refactor the entire script for RES
 * 1.5.2
   * Bug fix
 * 1.5.1
   * Bug fixes
   * Tweaked the preferences page styling
 * 1.5.0
   * Added a preferences page
   * Tip button does both public and private tipping
   * Switch to HTTPS for API
 * 1.4.5
   * Fix currency display precision
 * 1.4.4
   * Fix cancelled icon bug
   * Update to latest tip regexp
 * 1.4.3
   * Fix subreddit case-sensitivity bug
   * Adjusted "pending" icon color
 * 1.4.2
   * Bug fixes with balances
 * 1.4.0
   * Subreddit support indicator
   * Current balance display
   * New icons for richer status indicators
   * Support for reddit https (https everywhere).
 * 1.3.5
   * Drop JSONP for CORS to increase safety.
   * Attempt to link to source tip from tipped icons.
 * 1.3.4
   * Inject the bot status into r/bitcointip's sidebar
 * 1.3.3
   * Fix API URL.
   * Finer tip rejection control
   * Small bug fixes
 * 1.3.1
   * A tip button on posts
   * Verification icons on posts
   * Small display tweaks
 * 1.3.0
   * New verification and rejection markings
   * Use the new gettipped API
   * All new jQuery reddit plugin
 * 1.2.1
   * More efficient validation marking
   * Fixed status names
   * Run on all of reddit, including user pages
 * 1.2
   * Use the new bitcointip API to get status information
   * https://github.com/NerdfighterSean/bitcointip/blob/master/src/api/
 * 1.1.4
   * Don't hide verification when it has children
   * Don't hide verification when it's the URL target
 * 1.1.3
   * Fix case insensitivity in regex
 * 1.1.2
   * Display tipping amount
 * 1.1.1
   * Support the full tipping syntax
 * 1.1
   * Hide verification posts, inlining the verification
 * 1.0
   * Add a "tip bitcoins" button.


res.js
var RESUtils = {
    currentSubreddit: function() {
        var match = location.pathname.match(/\/r\/([^/]+)/);
        if (match) {
            return match[1].toLowerCase();
        } else {
            return null;
        }
    },
    loggedInUser: function() {
        var user = $('#header-bottom-right span.user a:first').text();
        return user === "login or register" ? null : user;
    },
    isMatchURL: function() {
        return true;
    },
    addCSS: function(css) {
        $(document.body).append($('<style/>').text(css));
    },
    /*
    Iterate through an array in chunks, executing a callback on each element.
    Each chunk is handled asynchronously from the others with a delay betwen each batch.
    If the provided callback returns false iteration will be halted.
    */
    forEachChunked: function(array, chunkSize, delay, call) {
        if (array == null) return;
        if (chunkSize == null || chunkSize < 1) return;
        if (delay == null || delay < 0) return;
        if (call == null) return;
        var counter = 0;
        var length = array.length;
        function doChunk() {
            for (var end = Math.min(array.length, counter+chunkSize); counter < end; counter++) {
                var ret = call(array[counter], counter, array);
                if (ret === false) return;
            }
            if (counter < array.length) {
                window.setTimeout(doChunk, delay);
            }
        }
        window.setTimeout(doChunk, delay);
    },
    notification: function(msg) {
        alert(msg.message);
    },
    commentsRegex: /https?:\/\/([a-z]+).reddit.com\/[-\w\.\/]*comments\/[-\w\.\/]*/i,
    friendsCommentsRegex: /https?:\/\/([a-z]+).reddit.com\/r\/friends\/*comments\/?/i,
    inboxRegex: /https?:\/\/([a-z]+).reddit.com\/message\/[-\w\.\/]*/i,
    profileRegex: /https?:\/\/([a-z]+).reddit.com\/user\/[-\w\.#=]*\/?(comments)?\/?(\?([a-z]+=[a-zA-Z0-9_%]*&?)*)?$/i, // fix to regex contributed by s_quark
    submitRegex: /https?:\/\/([a-z]+).reddit.com\/([-\w\.\/]*\/)?submit\/?$/i,
    prefsRegex: /https?:\/\/([a-z]+).reddit.com\/prefs\/?/i,
    commentPermalinkRegex: /comments\/[a-z0-9]+\/[^/]+\/[a-z0-9]+$/,
    pageType: function() {
        if (typeof(this.pageTypeSaved) == 'undefined') {
            var pageType = '';
            var currURL = location.href.split('#')[0];
            if (RESUtils.profileRegex.test(currURL)) {
                pageType = 'profile';
            } else if ((RESUtils.commentsRegex.test(currURL)) || (RESUtils.friendsCommentsRegex.test(currURL))) {
                pageType = 'comments';
            } else if (RESUtils.inboxRegex.test(currURL)) {
                pageType = 'inbox';
            } else if (RESUtils.submitRegex.test(currURL)) {
                pageType = 'submit';
            } else if (RESUtils.prefsRegex.test(currURL)) {
                pageType = 'prefs';
            } else {
                pageType = 'linklist';
            }
            this.pageTypeSaved = pageType;
        }
        return this.pageTypeSaved;
    },
    watchForElement: function() { /* stub */ },
    setCursorPosition: function(form, pos) {
        var elem = $(form)[0];
        if (!elem) return null;

        if (elem.setSelectionRange) {
            elem.setSelectionRange(pos, pos);
        } else if (elem.createTextRange) {
            var range = elem.createTextRange();
            range.collapse(true);
            range.moveEnd('character', pos);
            range.moveStart('character', pos);
            range.select();
        }

        return form;
    },
    isCommentPermalinkPage: function() {
        var path = document.location.pathname;
        return RESUtils.commentPermalinkRegex.test(path);
    },
    click: function(obj, button) {
        var evt = document.createEvent('MouseEvents');
        button = button || 0;
        evt.initMouseEvent('click', true, true, window.wrappedJSObject, 0, 1, 1, 1, 1, false, false, false, false, button, null);
        obj.dispatchEvent(evt);
    }
};

var RESConsole = {
    getModulePrefs: function () {
        return true;
    }
};

var RESStorage = {
    localStorage: (window.unsafeWindow || window).localStorage,
    setItem: function(key, value) {
        this.localStorage[key] = value;
    },
    getItem: function(key) {
        var value = this.localStorage[key];
        return value != null ? value : null;
    }
};

var modules = {};


footer.js
var bitcointip = modules['bitcointip'];
bitcointip.beforeLoad();
bitcointip.load();
if (!bitcointip.getAddress()) {
    bitcointip.fetchAddress();
}
bitcointip.go();

/* Create the options panel. */
if (location.pathname === '/r/bitcointip/wiki/preferences') {
    var $wiki = $('div.md.wiki');
    var $form = $('<form/>').attr({
        'class': 'tip-preferences roundfield'
    });

    /* Build up the form. */
    for (var item in bitcointip.options) {
        var option = bitcointip.options[item];
        var $label = $('<label/>').text(option.name).addClass('roundfield');
        var $input = null;
        if (option.type === 'text') {
            $input = $('<input/>').attr({
                name: item,
                type: 'text',
                value: option.value
            });
        } else if (option.type === 'boolean') {
            $input = $('<select/>').attr({
                name: item
            });
            var yes = $('<option/>').attr({
                value: 'true'
            }).text('yes');
            var no = $('<option/>').attr({
                value: 'false'
            }).text('no');
            if (option.value) {
                yes.attr('selected', true);
            } else {
                no.attr('selected', true);
            }
            $input.append(yes).append(no);
        } else if (option.type === 'enum') {
            $input = $('<select/>').attr({
                name: item
            });
            for (var i = 0; i < option.values.length; i++) {
                var value = option.values[i];
                var $option = $('<option/>').attr({
                    value: value.value
                }).text(value.value);
                if (option.value === value.value) {
                $option.attr('selected', true);
            }
                $input.append($option);
            }
        }
        if ($input) {
            var $div = $('<div/>');
            $div.append($label).append($input);
            $form.append($div);
        }
    }

    $form.append($('<input/>').attr({
        type: 'submit',
        value: 'Save Preferences'
    }));

    $form.submit(function(event) {
        event.preventDefault();
        for (var item in bitcointip.options) {
            var option = bitcointip.options[item];
            var $input = $('input[name=' + item + '], ' +
                           'select[name=' + item + ']');
            if ($input.length === 1) {
                console.log('Saved ' + item);
                var value = $input.val();
                if (option.type === 'boolean') {
                    value = JSON.parse(value);
                }
                bitcointip.options[item].value = value;
            } else {
                console.log('Skipped ' + item);
            }
        }
        bitcointip.save();
        window.location.reload(true); // Just to provide feedback
    });

    $wiki.empty().append($form);
}


/* ## Test Tips
 *
 * Rejected:
 *   http://www.reddit.com/r/bitcointip/comments/132nhq/_/c7c7iue
 * Rejected flip:
 *   http://www.reddit.com/r/Bitcoin/comments/14i9e7/_/c7dc6w9
 * Combination folding:
 *   http://www.reddit.com/r/bitcointip/comments/13iykn/_/c7dj8ia
 * Multiple tips to one receiver:
 *   http://www.reddit.com/r/bitcointip/comments/12lmut/_/c7ny177
 * Multiple guilded to one receiver (for comparison):
 *   http://www.reddit.com/r/AdviceAnimals/comments/15mk25/_/c7ntrcc
 * Reversed:
 *   http://www.reddit.com/r/IAmA/comments/18tp7t/_/c8i8qto
 * Fun tip to ignore:
 *   http://www.reddit.com/r/gaming/comments/18ihmm/_/c8f54l4
 */


bitcointip.user.js
// ==UserScript==
// @name         Extra Bitcointip Support on Reddit
// @description  Add a tipping button (/u/bitcointip) and shrinks verifications
// @version      1.5.2
// @license      Public Domain
// @include      http*://*.reddit.com/*
// @exclude      http*://*.reddit.com/user/bitcointip
// ==/UserScript==

/* Changelog:
 * 1.5.2
 *    Bug fix
 * 1.5.1
 *    Bug fixes
 *    Tweaked the preferences page styling
 * 1.5.0
 *    Added a preferences page
 *    Tip button does both public and private tipping
 *    Switch to HTTPS for API
 * 1.4.5
 *    Fix currency display precision
 * 1.4.4
 *    Fix cancelled icon bug
 *    Update to latest tip regexp
 * 1.4.3
 *    Fix subreddit case-sensitivity bug
 *    Adjusted "pending" icon color
 * 1.4.2
 *    Bug fixes with balances
 * 1.4.0
 *    Subreddit support indicator
 *    Current balance display
 *    New icons for richer status indicators
 *    Support for reddit https (https everywhere).
 * 1.3.5
 *    Drop JSONP for CORS to increase safety.
 *    Attempt to link to source tip from tipped icons.
 * 1.3.4
 *    Inject the bot status into r/bitcointip's sidebar
 * 1.3.3
 *    Fix API URL.
 *    Finer tip rejection control
 *    Small bug fixes
 * 1.3.1
 *    A tip button on posts
 *    Verification icons on posts
 *    Small display tweaks
 * 1.3.0
 *    New verification and rejection markings
 *    Use the new gettipped API
 *    All new jQuery reddit plugin
 * 1.2.1
 *    More efficient validation marking
 *    Fixed status names
 *    Run on all of reddit, including user pages
 * 1.2
 *    Use the new bitcointip API to get status information
 *    https://github.com/NerdfighterSean/bitcointip/blob/master/src/api/
 * 1.1.4
 *    Don't hide verification when it has children
 *    Don't hide verification when it's the URL target
 * 1.1.3
 *    Fix case insensitivity in regex
 * 1.1.2
 *    Display tipping amount
 * 1.1.1
 *    Support the full tipping syntax
 * 1.1
 *    Hide verification posts, inlining the verification
 * 1.0
 *    Add a "tip bitcoins" button.
 */

var tipregex = /\+(bitcointip|bitcoin|tip|btctip|bittip|btc)/i;
var tipregexFun = /(\+((?!0)(\d{1,4})) (point|internet|upcoin))/;
var botDownThreshold = 15 * 60 * 1000; // milliseconds
var botStatusHtml = {
    up: '<span class="status-up">UP</span>',
    down: '<span class="status-down">DOWN</span>'
};
var api = {
    gettips: 'https://bitcointip.net/api/gettips.php?',
    gettipped: 'https://bitcointip.net/api/gettipped.php?',
    subreddits: 'https://bitcointip.net/api/subreddits.php',
    balance: 'https://bitcointip.net/api/balance.php'
};
var icons = {
    completed: "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAA0AAAANCAMAAABFNRROAAAAt1BMVEX///8AAAAAyAAAuwAAwQcAvAcAvwAAwQYAyAUAxAUAxwQAwgQAvAMAxQYAvwYAxQYAxwU5yT060j460j871T89wUE9wkFGokdGu0hIzExJl09JmE9JxExJxE1K1U9K1k5Ll09LmVNMmVNM2FBNmlRRx1NSzlRTqlVUslZU1ldVq1hVrFdV2FhWrFhX21pZqlphrWJh3WRotGtrqm1stW91sXd2t3h5t3urz6zA2sHA28HG3sf4+PhvgZhQAAAAEXRSTlMAARweJSYoLTM0O0dMU1dYbkVIv+oAAACKSURBVHjaVc7XEoIwEIXhFRED1tBUxBaPFSyxK3n/5zIBb/yv9pudnVky2Ywxm345MHkVXByllPm4W24qrLbzdo1sLPPRepc+XlnSIAuz9DQYPtXnkLhUF/ysrndV3CYLRpbg2VtpxFMwfRfEl8IghEPUhB9t9lEQoke6FnzONfpU5kEIoKOn/z+/pREPWTic38sAAAAASUVORK5CYII=",
    cancelled: "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAA0AAAANCAMAAABFNRROAAAAQlBMVEX///+qAAAAAAC/AADIABSaTU3YMDDcPj7cSEjeUFDiZGTld3fmfHzoiorqkJDqlpbupKTuqqr99fX99/f+/Pz////kWqLlAAAABXRSTlMAAwcIM6KYVMQAAABfSURBVHjaXc7JDsAgCEVRsYpIBzro//9qHyHpond3AgkklIuXPKJcqleIIEB6FwEhQEW0t4rlUtt+22ZTMQ09NqZyiK8BtBCvc9iDWegY526hBVRmdcQ9RgD9f/G+P1+JEwRF2vKhRgAAAABJRU5ErkJggg==",
    tipped: "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAwAAAANCAMAAACq939wAAAA/FBMVEWqVQCqcQCZZgCLXQCdYgDMjACOXACOXwCacQCfcQCqegCwggChggCnfwCwggCthACtgQC9iACleQC+iwDMlgDJlACmfgDDiwDBjADHlQDJnQDFlQDKmAChfRGjgBOmfhKmghKnghOqhBKthxOviROvjB+vjCGvjCOwiRSwihixixWxjSGziBOzkSmzky+0kSa0kiy3jxW8kxS8mCi8n0i8oEq9lBe9mSvOoBbUrTTUrTjbukXcsTDctDrdtj/exHbexXDfwWXfwmvjrBnksRjksx7lrhnqx17qyWTqymrq377rz3nr2qftuiHtvSv67cD67sj+997/+OX///8rcy1sAAAAHXRSTlMDCQoLDRQkKystMDc5Oj0+QUlKS0tMTVFSV15/i6wTI/gAAACWSURBVHjaHcrnAoFQGADQryI7sjNKN0RKZJVNQ8io938YN+f3ASDp1B9NAhD15UzXNH26KhNQXZyDBxZcNlloKadvFIbR56owUOFV9425ai8PrGwZaITQeisXoKHs7k/MP44ZaHYl54U5El8EdmjNEWbEjesf/Ljd9v0S1ETBtD3PNgUxB8nOQIybOGknAKgMy2FsmoIflIEZdK7PshkAAAAASUVORK5CYII=",
    pending: "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAA0AAAANCAYAAABy6+R8AAABWUlEQVR4nI2Sy0sCURTGD6S2jjYFrdy0DNpEhokb8zFm5YyaO6NFYNGqF/0hPZYtR79FUbgw0BFtDKIgUCSpv8Od3XtGJzWDBj64h/l+954XdbtdGhQZkzNUd7ptifiXZygo0Wz0WsWoyHTMj4Wo6nRLQ7KdRuZz15bWSiF0GQOVXJ4hqP/COGDTjEO9SyByIcDHiUXiT+QsAaW1wabgi4KtVxVqM4lQVcFx4RS5tzy0vIgZFDVTnaYkFG6us2lbTyNws4ZAMYizwjk6nQ7KbQOJfMqCRBlERZpWruJYfvYigx02ZfUDHN2e8Pnpy8T+w6G4MIqI8HFH5Ut9SKZQ/jDYPAh4K36EGzGrkwz1avK8+/jn3n2WzaPASsNnQaJpvYG65ixwFV7Dj7iuQcul+Cwvs4Ga1fafOVUcC31Qpio1BJjO0PiNEJPn9osapeyNqLmW/lyj/+7eN1qRZT0kKLSqAAAAAElFTkSuQmCC",
    reversed: "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAA0AAAANCAYAAABy6+R8AAABOklEQVR42p2SvU4CQRSFj+9h7DQ2VmsDCy0/Cw2wsNtuRwixIiQ8CZaWGDBaaAiJyVAg2xi1AaTQ19jyOHeSIbLBxuIkM7Pn23PvnQHJPSngNAYcK9mnPWng/LpaZVoadg9CC+BSDNsg4FcU7bRpNjkslaiAYAfZhL+AuFbjQ6PBYaXCZ6AK4Mj0IMBGH4rptVjkW73Ote9zUS5z2u/zfTzmRBL1XoNniIFjgdaeZ0yjMORNocCZ1nQwYJIk3CrFSatlIGkDM+BEouNMhndRZEyjTof3vZ5Zfy+XfOp2zQ+ND3BMkoWkhE+lxLwHzPN5rjzPTtLZ9fSRzZqPj+22MaeBlesaSIZmp3dhQZXLcaQHcev7sjZnFngBwr17mgNFC+pSRRawZV0dfBFy89ogDYvEbBMav33/ens/XHaDp7U/bFsAAAAASUVORK5CYII="
};
var displayCurrency = {
    balanceUSD: {unit: '$', precision: 2},
    balanceBTC: {unit: '฿'},
    balanceJPY: {unit: '¥'},
    balanceGBP: {unit: '£', precision: 2},
    balanceEUR: {unit: '€', precision: 2}
};
var $ = unsafeWindow.$,
    S = unsafeWindow.localStorage,
    reddit = unsafeWindow;

/* Helper functions. */

function identity(x) {
    return x;
}

/**
 * Set textarea cursor position in jQuery.
 */
$.fn.setCursorPosition = function(pos) {
    this.each(function(index, elem) {
        if (elem.setSelectionRange) {
            elem.setSelectionRange(pos, pos);
        } else if (elem.createTextRange) {
            var range = elem.createTextRange();
            range.collapse(true);
            range.moveEnd('character', pos);
            range.moveStart('character', pos);
            range.select();
        }
    });
    return this;
};

function quantity(object) {
    var pref = preferences.currency.toUpperCase();
    var unit = displayCurrency['balance' + pref];
    var amount = object['amount' + pref];
    return unit.unit + amount;
}

/* Preferences object. */
var preferences = null;

function savePreferences() {
    S.bitcointipPreferences = JSON.stringify(preferences);
}

function loadPreferences() {
    try {
        preferences = JSON.parse(S.bitcointipPreferences);
    } catch(e) {
        preferences = null;
    }
    if (preferences == null) {
        preferences = {
            baseTip: '0.01 BTC',
            hide: true,
            status: 'detailed',
            currency: 'BTC',
            balance: true,
            subreddit: true
        };
        savePreferences();
    }
    if (preferences.status === 'basic') {
        icons.pending = icons.completed;
        icons.reversed = icons.completed;
    }
}

loadPreferences();

/* Add the "tip bitcoins" button after "give gold". */
var tip =
  $('<span class="tip-wrapper">' +
      '<div class="dropdown" onclick="open_menu(this)">' +
        '<a class="tip-bitcoins login-required">tip bitcoins</a>' +
      '</div>' +
      '<div class="drop-choices">' +
        '<a class="choice tip-publicly">tip publicly</a>' +
        '<a class="choice tip-privately">tip privately</a>' +
      '</div>' +
    '</span>');

if (/^\/r\//.test(document.location.pathname)) {
    $('a.give-gold').parent().after($('<li/>').append(tip.clone()));
    if ($('.link').length === 1) { // Viewing a submission?
        $('.link ul.buttons').append($('<li/>').append(tip.clone()));
    }
}

$('.tip-publicly').click(function(event) {
    event.preventDefault();
    var $target = $(event.target);
    var form = null;
    if ($target.closest('.link').length > 0) { /* Post */
        form = $('.commentarea .usertext:first');
    } else { /* Comment */
        reddit.reply(event.target);
        form = reddit.comment_reply_for_elem(event.target);
    }
    var textarea = form.find('textarea');
    if (!textarea.val().match(tipregex)) {
        textarea.val(textarea.val() + '\n\n+tip ' + preferences.baseTip);
        textarea.setCursorPosition(0);
    }
});

$('.tip-privately').click(function(event) {
    event.preventDefault();
    var $target = $(event.target);
    var form = null;
    if ($target.closest('.link').length > 0) { /* Post */
        form = $('.commentarea .usertext:first');
    } else {
        form = $target.thing().find(".child .usertext:first");
    }
    if (form.length > 0 && form.find('textarea').val()) {
        /* Confirm if a comment has been entered. */
        if (!confirm('Really leave this page to tip privately?')) {
            return;
        }
    }
    var user = $(this).thing().find('.author').first().text();
    var message = encodeURIComponent('+bitcointip @' + user + ' ' +
                                     preferences.baseTip);
    var url = '/message/compose?to=bitcointip&subject=Tip&message=' + message;
    window.location = url;
});

/* Subreddit indicator. */
var subreddit = (function(match) {
    if (match) {
        return match[1];
    } else {
        return null;
    }
}(location.pathname.match(/\/r\/([^/]+)/)));
if (preferences.subreddit && subreddit) {
    $.getJSON(api.subreddits, function(data) {
        if (data.subreddits.indexOf(subreddit.toLowerCase()) >= 0) {
            $('#header-bottom-right form.logout')
                .before($('<span>|</span>').attr({
                    'class': 'separator'
                }))
                .prev().before($('<img/>').attr({
                    'src': icons.tipped,
                    'style': 'vertical-align: text-bottom;',
                    'title': 'Tips enabled in this subreddit.'
                }));
        }
    });
}

/* Balance indicator. */
var user = $('#header-bottom-right span.user a').first().text();
if (user === "login or register") {
    user = null;
}
var address = S['address.' + user];
if (S.balanceUnits == null) {
    S.balanceUnits = 'balanceUSD';
}

function toggleBalanceUnits() {
    var units = Object.keys(displayCurrency);
    var i = (units.indexOf(S.balanceUnits) + 1) % units.length;
    return (S.balanceUnits = units[i]);
}

function currencyString(balance) {
    var meta = displayCurrency[S.balanceUnits];
    var quantity = balance[S.balanceUnits];
    if (meta.precision) {
        quantity = quantity.toFixed(meta.precision);
    }
    return meta.unit + quantity;
}

function insertBalance() {
    $.getJSON(api.balance, {
        username: user,
        address: address
    }, function (balance) {
        if (!('balanceBTC' in balance)) {
            /* Probably got the address wrong. */
            S.removeItem('address.' + user);
            return;
        }
        var units = S.balanceUnits;
        $('#header-bottom-right form.logout').before($('<span>|</span>').attr({
            'class': 'separator'
        })).prev().before($('<a/>').attr({
            'class': 'hover',
            'href': '#'
        }).on('click', function() {
            toggleBalanceUnits();
            $(this).text(currencyString(balance));
        }).text(currencyString(balance)));
    });
}

if (preferences.balance && user != null && address == null) {
    $.getJSON('/message/messages.json', function(messages) {
        /* Search messages for a bitcointip response. */
        address = messages.data.children.filter(function (message) {
            return message.data.author === 'bitcointip';
        }).map(function (message) {
            var pattern = /Deposit Address: \| \[\*\*([a-zA-Z0-9]+)\*\*\]/;
            var address = message.data.body.match(pattern);
            if (address) {
                return address[1];
            } else {
                return false;
            }
        }).filter(identity)[0]; // Use the most recent
        if (address) {
            S['address.' + user] = address;
            insertBalance();
        }
    });
} else if (preferences.balance && user != null && address != null) {
    insertBalance();
}

/* Reddit jQuery plugin. */
(function($) {
    /** Get the comment div for each element in the current set. */
    $.fn.comment = function() {
        return this.closest('.comment');
    };

    /** Get the comment ID for the first selected comment. */
    $.fn.commentID = function() {
        var full = this.first().find('input[name="thing_id"]').first().val();
        return full.replace(/^t1_/, '');
    };

    /** Get the commenter/poster name for the first selected comment. */
    $.fn.thingName = function() {
        return this.first().find('.author').first().text();
    };

    /** Get the comment's post time for the first selected comment. */
    $.fn.commentDate = function() {
        return new Date(this.find('.tagline time').first().attr('datetime'));
    };

    /** Get the comment body for the first comment in the current set. */
    $.fn.commentBody = function() {
        return this.find('.md').first();
    };

    /** Get the children comments for each comment in the current set. */
    $.fn.commentChildren = function() {
        return this.find('.comment');
    };

    /** Get the parent comment for each comment in the current set. */
    $.fn.commentParent = function() {
        return this.parent().comment();
    };

    /** Determine whether the first element is the target comment. */
    $.fn.isTarget = function() {
        return this.first().find('form').first().hasClass('border');
    };

    /** Return true if the first comment in the current set has a tip. */
    $.fn.hasTip = function() {
        return this.commentBody().children().is(function() {
            return tipregex.test($(this).text());
        });
    };

    /** Return true if the first comment in the current set has a "fun" tip. */
    $.fn.hasFunTip = function() {
        return this.commentBody().children().is(function() {
            return tipregexFun.test($(this).text());
        });
    };

    /** Return the link ID for the first post in the selection. */
    $.fn.postID = function() {
        return this.attr('data-fullname').replace(/^t._/, '');
    };

    /** Return true if the first seleted item is a comment. */
    $.fn.isComment = function() {
        return this.closest('.link').length === 0;
    };
})(unsafeWindow.jQuery);

/* Hide verification replies. Note: t2_7vw3n is /u/bitcointip. */
if (preferences.hide) {
    $('a.id-t2_7vw3n').comment().each(function() {
        var $this = $(this);
        if ($this.commentChildren().length === 0 && !$this.isTarget()) {
            reddit.hidecomment($this.find('.expand').first());
        }
    });
}

/* Find all the tip comments. */
var tips = {};
var fun = {};
$('div.comment').each(function() {
    var $this = $(this);
    if ($this.hasTip()) {
        tips[$this.commentID()] = $this;
    } else if ($this.hasFunTip()) {
        tips[$this.commentID()] = $this;
        fun[$this.commentID()] = $this;
    }
});

/* Get status information about various tips. */
var inTipSubreddit = /^\/r\/bitcointip/.test(document.location.pathname);
var tipIDs = Object.keys(tips);
var confirmedIDs = [];
if (preferences.status !== 'none' && (tipIDs.length > 0 || inTipSubreddit)) {
    var iconStyle = 'vertical-align: text-bottom; margin-left: 8px;';
    $.getJSON(api.gettips + 'tips=' + tipIDs, function(response) {
        var lastEvaluated = new Date(response.last_evaluated * 1000);
        if (inTipSubreddit) {
            var botStatus = null;
            if (Date.now() - lastEvaluated > botDownThreshold) {
                botStatus = botStatusHtml.down;
            } else {
                botStatus = botStatusHtml.up;
            }
            $('a[href="http://bitcointip.net/status.php"]').html(botStatus);
        }

        response.tips.forEach(function (tip) {
            var id = tip.fullname.replace(/^t._/, '');
            var tagline = tips[id].find('.tagline').first();
            var icon = $('<a/>').attr({href: tip.tx, target: '_blank'});
            tagline.append(icon.append($('<img/>').attr({
                src: icons[tip.status],
                style: iconStyle,
                title: quantity(tip) + ' → ' + tip.receiver +
                    ' (' + tip.status + ')'
            })));
            confirmedIDs.push(id);
            tips[id].attr('id', 't1_' + id);
            delete tips[id];
        });

        /* Deal with unanswered tips. */
        for (var id in tips) {
            if (!fun[id] && tips[id].commentDate() < lastEvaluated) {
                var tagline = tips[id].find('.tagline').first();
                tagline.append($('<img/>').attr({
                    src: icons.cancelled,
                    style: iconStyle,
                    title: 'This tip is invalid.'
                }));
            }
        }
    });

    /* Put receiver information on comments. */
    var things = {};
    $('div.comment').each(function() {
        var $this = $(this);
        things[$this.commentID()] = $this;
    });
    $('div.link').each(function() {
        var $this = $(this);
        things[$this.postID()] = $this;
    });
    var thingIDs = Object.keys(things);
    $.getJSON(api.gettipped + 'tipped=' + thingIDs, function(response) {
        response.forEach(function (tipped) {
            var id = tipped.fullname.replace(/^t._/, '');
            var thing = things[id];
            var tagline = thing.find('.tagline').first();
            var plural = tipped.tipQTY > 1;
            var title = quantity(tipped) + ' to ' + thing.thingName() +
                    ' for this ';
            if (plural) {
                title = 'redditors have given ' + title;
            } else {
                title = 'a redditor has given ' + title;
            }
            if (thing.isComment()) {
                title += 'comment.';
            } else {
                title += 'submission.';
            }
            var icon = $('<img/>').attr({
                src: icons.tipped,
                style: iconStyle,
                title: title
            });

            /* Attempt to link to the first tip. */
            var foundParent = false;
            thing.commentChildren().each(function() {
                if (!foundParent) {
                    var $this = $(this);
                    var id = $this.commentID();
                    if ($.inArray(id, confirmedIDs) >= 0) {
                        icon = $('<a/>').attr({
                            href: '#t1_' + $this.commentID()
                        }).append(icon);
                        foundParent = true;
                    }
                }
            });

            tagline.append(icon);
            if (plural) {
                tagline.append($('<span/>').text('x' + tipped.tipQTY));
            }
        });
    });
}

/* Preferences */
(function() {
    if (location.pathname !== '/r/bitcointip/wiki/preferences') return;
    $('div.md.wiki').wrap($('<form/>').attr({
        'class': 'tip-preferences roundfield'
    }));

    function makeText(name, value) {
        return $('<input/>').attr({
            name: name,
            type: 'text',
            value: value,
            'class': 'tip-preference'
        });
    }

    function makeSelect(name, options, def) {
        if (def == null) {
            def = preferences[name];
            if (typeof def === 'boolean') {
                def = options[def ? 0 : 1].toLowerCase();
            }
        }
        var $input = $('<select/>').attr({
            name: name,
            'class': 'tip-preference'
        });
        for (var i = 0; i < options.length; i++) {
            var v = options[i];
            var $option = $('<option/>').attr({
                value: v.toLowerCase()
            }).text(v);
            if (def === v.toLowerCase()) {
                $option.attr('selected', true);
            }
            $input.append($option);
        }
        return $input;
    }

    function fix(name, input) {
        var $li = $('a[href="/pref-' + name + '"]').parent();
        var label = $li.text();
        $li.empty().append($('<label/>').text(label).addClass('roundfield'))
            .append(input);
    }

    /* Build up the form. */
    fix('basetip', makeText('basetip', preferences.baseTip));
    fix('hide', makeSelect('hide', ['Yes', 'No']));
    fix('status', makeSelect('status', ['Detailed', 'Basic', 'None']));
    fix('currency', makeSelect('currency', ['USD', 'BTC']));
    fix('balance', makeSelect('balance', ['Yes', 'No']));
    fix('subreddit', makeSelect('subreddit', ['Yes', 'No']));

    var $submitA = $('a[href="/pref-submit"]');
    $submitA.replaceWith($('<input/>').attr({
        type: 'submit',
        value: $submitA.text() || 'Save Preferences'
    }));

    var $form = $('form.tip-preferences');
    function v(name) {
        return $form.find('.tip-preference[name="' + name + '"]').val();
    }

    $('form.tip-preferences').submit(function(event) {
        event.preventDefault();
        preferences.baseTip = v('basetip');
        preferences.hide = v('hide') === 'yes';
        preferences.status = v('status');
        preferences.currency = v('currency');
        preferences.balance = v('balance') === 'yes';
        preferences.subreddit = v('subreddit') === 'yes';
        savePreferences();
        window.location.reload(true); // Just to provide feedback
    });
}());

/* Test URLs:
 *
 * Rejected,
 *   http://www.reddit.com/r/bitcointip/comments/132nhq/t/c7c7iue
 *
 * Rejected flip,
 *   http://www.reddit.com/r/Bitcoin/comments/14i9e7/y/c7dc6w9
 *
 * Combination folding,
 *   http://www.reddit.com/r/bitcointip/comments/13iykn/b/c7dj8ia
 *
 * Multiple tips to one receiver,
 *   http://www.reddit.com/r/bitcointip/comments/12lmut/c7ny177
 *
 * Multiple guilded to one receiver (for comparison),
 *   http://www.reddit.com/r/AdviceAnimals/comments/15mk25/c7ntrcc
 *
 * Reversed,
 *   http://www.reddit.com/r/IAmA/comments/18tp7t/c8i8qto
 */
